### PR TITLE
Add is_leaf argument to pytrees\pytrees.py with test at tests\pytrees\test_pytrees.py, fixes #6083

### LINKED ISSUE #6083: Add `is_leaf`optional keyword argument to `qml.pytrees.flatten`
PennyLane has its own tools for working with pytrees similar to jax's [`tree`](https://jax.readthedocs.io/en/latest/jax.tree.html#module-jax.tree) module.  These tools can be used to manipulate data structures.

```pycon
>>> x = {"a": (1, 2), "b": 3}
>>> data, struct = qml.pytrees.flatten(x)
>>> data
[1, 2, 3]
>>> struct
PyTreeStructure(dict, ('a', 'b'), [PyTreeStructure(tuple, None, [PyTreeStructure(), PyTreeStructure()]), PyTreeStructure()])
>>> qml.pytrees.unflatten([4,5,6], struct)
{'a': (4, 5), 'b': 6}
```

These utilities are located in [`pennylane/pytrees/pytrees.py/`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pytrees/pytrees.py)

This task is to add a `is_leaf: Optional[Callable[[Any], bool]] = None` similar to that in [`jax.tree.flatten`](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree.flatten.html#jax.tree.flatten).

If provided and returning `True`, this callable should take precedence over whether or not the type has a `flatten_fn` for determining if the object is a terminal leaf.

For example,
```
>>> def is_leaf(x: Any) -> bool:
...     return isinstance(x, tuple)
>>> data, struct = qml.pytrees.flatten(x, is_leaf=is_leaf)
>>> data
[(1, 2), 3]
>>> struct
PyTreeStructure(dict, ('a', 'b'), [PyTreeStructure(), PyTreeStructure()])
```

Note that `(1,2)` is now considered a terminal leaf.

A PR should:

* Tests in [`tests/pytrees/test_pytrees.py`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/pytrees/test_pytrees.py)
*  A changelog entry
* Obey black, pylint, and isort

### PR DESCRIPTION
**Pull request by tonmoy-b**
**Context:**
Add is_leaf argument to pytrees\pytrees.py with test at tests\pytrees\test_pytrees.py
fixes #6083 

**Description of the Change:**
Added is_leaf argument to the function pytrees\pytrees.py. Should this argument evaluate to True then no fufther flattening will be done. 
Added test to tests\pytrees\test_pytrees.py, with function test_flatten_is_leaf() having two test cases.
Added changelog with file \doc\releases\changelog-0.38.0.md

**Benefits:**
argument added to the function flatten which acts as a control mechanism to designate a node as leaf node even if that node could be further flattened, e.g., the node is a tuple.

**Possible Drawbacks:**

**Related GitHub Issues:**
Fixes #6083 

**Related Shortcut Stories:**
[sc-70903]

### CODE REVIEW

File: pennylane/pytrees/pytrees.py [hunk 1]
```diff
@@ -217,11 +217,18 @@ def __str__(self):
 leaf = PyTreeStructure(None, (), [])
 
 
-def flatten(obj: Any) -> tuple[list[Any], PyTreeStructure]:
+def flatten(
+    obj: Any, is_leaf: Callable[[Any], bool] | None = None
```
Comment (albi3ro) [inline pennylane/pytrees/pytrees.py:221]:
```suggestion
    obj: Any, is_leaf: Optional[Callable[[Any], bool]] = None
```
Note that we still support Python 3.9, at least for the near future.  The `| None` syntax is only available in Python 3.10+.

**Review by albi3ro — COMMENTED** [review]
Thanks for opening this PR :) Looks great other than the 3.10+ typing hinting syntax.

File: pennylane/pytrees/pytrees.py [hunk 2]
```diff
@@ -217,11 +217,18 @@ def __str__(self):
...
+) -> tuple[list[Any], PyTreeStructure]:
     """Flattens a pytree into leaves and a structure.
 
     Args:
         obj (Any): any object
+        is_leaf (Callable[[Any], bool] | None = None): an optionally specified
+        function that will be called at each flattening step. It should return
+        a boolean, with true stopping the traversal and the whole subtree being
+        treated as a leaf, and false indicating the flattening should traverse
+        the current object.
```
Comment (mudit2812) [inline pennylane/pytrees/pytrees.py:231]:
```suggestion
        is_leaf (Callable[[Any], bool] | None = None): an optionally specified
            function that will be called at each flattening step. It should return
            a boolean, with ``True`` stopping the traversal and the whole subtree being
            treated as a leaf, and ``False`` indicating the flattening should traverse
            the current object.
```

Comment (mudit2812) [reply pennylane/pytrees/pytrees.py:231]: The indentation is needed for the documentation to render properly.

**Comment by tonmoy-b** [comment]
hi @albi3ro and @mudit2812, thanks for the review and suggestions. I have accepted the suggested changes. Do let me know if it would be nice for me to look work on any other contributions, I'd love to get more familiar with the code.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -197,6 +197,9 @@
   relaxation to occur during longer measurements. The trit flip error represents misclassification during readout.
   [(#5842)](https://github.com/PennyLaneAI/pennylane/pull/5842)
 
+* Added `is_leaf` parameter to function `flatten` in `pytrees.pytrees.py`. This is to allow node flattening to be stopped for any node where the `is_leaf` Optional argument evaluates to being true.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:200]:
```suggestion
* Added `is_leaf` parameter to function `flatten` in the `qml.pytrees` module. This is to allow node flattening to be stopped for any node where the `is_leaf` optional argument evaluates to being `True`.
```

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -197,6 +197,9 @@
...
+  [(#6083)](https://github.com/PennyLaneAI/pennylane/issues/6083)
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:201]: The link should be to the PR, not to the issue 😄 

```suggestion
  [(#6107)](https://github.com/PennyLaneAI/pennylane/issues/6107)
```

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -338,6 +341,7 @@ Tarun Kumar Allamsetty,
 Guillermo Alonso,
 Utkarsh Azad,
 Gabriel Bottrill,
+Tonmoy T. Bhattacharya,
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:344]:
```suggestion
Tonmoy T. Bhattacharya,
Gabriel Bottrill,
```

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks for opening this PR @tonmoy-b ! The implementation looks great. I left some suggestions for the changelog. Do you mind also adding a couple of more tests:
* A test to verify that the pytree structure returned by `flatten` is correct when given an `is_leaf` argument.
* A test to verify that flattening and then unflattening the same pytree will be equivalent to the original object when `flatten` is given an `is_leaf` argument.

Other than the above comments, the PR is in great shape and I'm happy to approve once the above comments are resolved 🎉

**Review by albi3ro — APPROVED** [review]
👍 Thanks for this PR :)

**Comment by Alex-Preciado** [comment]
Thank you so much for this PR @tonmoy-b 🚀 !! We will continue with the review next week. Have a great weekend!!

**Comment by tonmoy-b** [comment]
> Thanks for opening this PR @tonmoy-b ! The implementation looks great. I left some suggestions for the changelog. Do you mind also adding a couple of more tests:
> 
> * A test to verify that the pytree structure returned by `flatten` is correct when given an `is_leaf` argument.
> * A test to verify that flattening and then unflattening the same pytree will be equivalent to the original object when `flatten` is given an `is_leaf` argument.
> 
> Other than the above comments, the PR is in great shape and I'm happy to approve once the above comments are resolved 🎉

Hi @mudit2812, @albi3ro  @Alex-Preciado , 

      I have added the mentioned tests into tests\pytrees\test_pytrees.py, as three functions (test_flatten_is_leaf, test_unflatten_is_leaf, test_unflatten_for_structure_is_leaf), kindly take a look. 

Regards

**Review by mudit2812 — APPROVED** [review]
Thanks for adding the suggested tests! Very happy to get this in 😄

**Comment by Alex-Preciado** [comment]
Thank you so much for this contribution @tonmoy-b !! 🚀

**Comment by tonmoy-b** [comment]
> Thank you so much for this contribution @tonmoy-b !! 🚀 

Thanks Alex, it was my pleasure and I hope to make more contributions to Pennylane going forward.

**Pull request merged**